semanticdb3/semanticdb3.md

```diff
@@ -0,0 +1,491 @@
+two unpleasant choices. First, they can use compiler internals, which are often
+underdocumented and lack compatibility guarantees. Otherwise, they reimplement
+compiler internals, which usually leads to duplication of effort and inconsistent UX.
+For example, Scala IDE [\[9\]][9] uses Scala compiler internals, which has known
+stability problems in interactive mode. To the contrast, IntelliJ [\[10\]][10] has
+its own Scala typechecker, which is more stable but is known for spurious red squiggles.
+
+This demonstrates the necessity for portable metaprogramming APIs -
+something that we have been working on within Scalameta [\[11\]][11].
+In the previous years, we shipped portable syntactic APIs for Scala,
+including abstract syntax trees, parsing and prettyprinting [\[12\]][12].
+SemanticDB is our take on portable semantic APIs.
+
+## Overview
+
+SemanticDB is a data model for semantic information about programs in Scala and
+other languages. SemanticDB decouples production and consumption of semantic information,
+establishing documented means for communication between tools.
+
+TODO: Once we upgrade semanticdb-scalac to support SemanticDB v3,
+demonstrate a hello world example. Show the original program, a prettyprinted
+version of its SemanticDB data, very briefly explain the sections.
+
+TODO: Once we ship metac and metap that support SemanticDB v3, explain that
+the SemanticDB data model is materialized as Protocol Buffers [\[13\]][13], JSON [\[14\]][14]
+and SQL [\[15\]][15]. Demonstrate how we can use metac to generate `.semanticdb` files
+from the hello world example above, show the binary payload, decode the files with metap.
+
+TODO: Explain how these protobuf files can be used in tools like scalafix, metadoc and metals.
+Mention that beyond protobuf files, there are many ways to use SemanticDB as an
+interchange format, including passing objects around in memory, communication via JSON, etc etc.
+
+## Model
+
+In this section, we describe the SemanticDB data model by going through the
+individual sections of the associated protobuf schema. However, in addition to protobuf,
+we also support other kinds of schemas, including JSON and SQL.
+See [Schemas](#schemas) for more information.
+
+### TextDocument
+
+```protobuf
+message TextDocuments {
+  repeated TextDocument textDocuments = 1;
+}
+
+message TextDocument {
+  string model = 255;
+  string uri = 1;
+  string text = 2;
+  string language = 3;
+  repeated Identifier identifiers = 4;
+  repeated SymbolInformation symbols = 5;
+  repeated Diagnostic diagnostics = 6;
+  repeated Synthetic synthetics = 7;
+}
+```
+
+`TextDocument` is the central data structure in the SemanticDB model.
+It provides semantic information about a snippet of code written in a programming
+language `language`.
+
+SemanticDB payloads must include the descriptor of the data model that was used
+to produce them in the `model` field. The following descriptors are supported:
+
+<table>
```

Comment:
Why write manual table instead of markdown table?

Reply:
I find tables easier to maintain than the markdown ASCII art, which doesn't give control of column widths and breaks down on non-trivial formatting. If you'd like, I can try to migrate to the Markdown style.

Reply:
Interesting, I agree maintaining markdown tables is a drag. Sure, let's do this 👍